### PR TITLE
feat: Native filters fast-follow

### PR DESCRIPTION
**Pull request by rusackas**
### SUMMARY

Adds onto https://github.com/apache/incubator-superset/pull/11814

This branch isn't my own work so much as a party! This is a feature branch containing the excellent work of:
@suddjian
@villebro 
@pkdotson 
@agatapst
@simchaNielsen
@ amitmiran137

NEW STUFF! 🎉
• adjust styling to match design spec
• better filter delete UX
• cascading filters
• error boundaries
• non-instant filtering
• default values

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Screen Shot 2020-12-18 at 11 54 18 PM
Screen Shot 2020-12-18 at 11 53 36 PM
Screen Shot 2020-12-18 at 11 53 22 PM
Screen Shot 2020-12-18 at 11 53 09 PM

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/spec/javascripts/utils/cacheWrapper_spec.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { cacheWrapper } from 'src/utils/cacheWrapper';
+
+describe('cacheWrapper', () => {
```
Comment (suddjian) [inline superset-frontend/spec/javascripts/utils/cacheWrapper_spec.ts:21]: nice tests

File: superset-frontend/src/dashboard/components/nativeFilters/CascadePopover.tsx [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  onExtraFormDataChange,
+}) => {
+  const getActiveChildren = useCallback((filter: CascadeFilter):
+    | CascadeFilter[]
+    | null => {
+    const children = filter.cascadeChildren || [];
+    const currentValue = filter.currentValue || [];
+
+    const activeChildren = children.flatMap(
+      childFilter => getActiveChildren(childFilter) || [],
```
Comment (suddjian) [inline superset-frontend/src/dashboard/components/nativeFilters/CascadePopover.tsx:83]: cool use of `flatMap`

File: superset-frontend/src/dashboard/components/nativeFilters/CascadePopover.tsx [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      <CascadeFilterControl
+        data-test="cascade-filters-control"
+        key={filter.id}
+        filter={filter}
+        onExtraFormDataChange={onExtraFormDataChange}
+      />
+    </>
+  );
+
+  const activeFilters = getActiveChildren(filter) || [filter];
```
Comment (suddjian) [inline superset-frontend/src/dashboard/components/nativeFilters/CascadePopover.tsx:137]: Why are we calling these "active" filters? What exactly does that mean here?

File: superset-frontend/src/dashboard/components/nativeFilters/FilterConfigModal.tsx [hunk 4]
```diff
@@ -149,9 +276,68 @@ export function FilterConfigModal({
     );
   }
 
+  function getParentFilters(id: string) {
+    return filterIds
+      .filter(filterId => filterId !== id && !removedFilters[filterId])
```
Comment (suddjian) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterConfigModal.tsx:281]: Good catch keeping removed filters out

**Review by suddjian — APPROVED** [review]
The parts I didn't write LGTM!

**Comment by junlincc** [comment]
if the code look good, let's merge first? we should test it we all other recent merged PRs

File: superset-frontend/src/dashboard/components/nativeFilters/CascadePopover.tsx [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  onExtraFormDataChange: (filter: Filter, extraFormData: ExtraFormData) => void;
+}
+
+const StyledTitleBox = styled.div`
+  display: flex;
+  flex-direction: row;
+  justify-content: space-between;
+  align-items: center;
+  background-color: ${({ theme }) => theme.colors.grayscale.light4};
+  margin: -5px -16px; // to override default antd padding
```
Comment (rusackas) [inline superset-frontend/src/dashboard/components/nativeFilters/CascadePopover.tsx:40]:
```suggestion
  margin: ${({ theme }) => theme.gridUnit * -1}px; ${({ theme }) => theme.gridUnit * -4}px; // to override default antd padding
```
Hopefully -4px is close enough to -5px?

**Comment by rusackas** [comment]
Reviewed the code and it LGTM!

File: superset-frontend/src/utils/cacheWrapper.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+export const cacheWrapper = <T extends Array<any>, U>(
```
Comment (rusackas) [inline superset-frontend/src/utils/cacheWrapper.ts:20]: Not sure how much sleep we should lose over the "any"s in this area.

Comment (suddjian) [reply superset-frontend/src/utils/cacheWrapper.ts:20]: I tried removing it and it broke stuff, I think it needs to stay

**Pull request merged**